egui/src/widgets/plot/legend.rs

```diff
@@ -1,81 +1,231 @@
         let (rect, response) = ui.allocate_exact_size(desired_size, Sense::click());
-        let rect = rect.shrink2(padding);
 
         response.widget_info(|| WidgetInfo::selected(WidgetType::Checkbox, *checked, &galley.text));
 
         let visuals = ui.style().interact(&response);
+        let flipped = ui.layout().cross_align() == Align::RIGHT;
 
-        let (small_icon_rect, big_icon_rect) = ui.spacing().icon_rectangles(rect);
+        let icon_position_x = if flipped {
+            rect.right() - icon_size / 2.0
+        } else {
+            rect.left() + icon_size / 2.0
+        };
+        let icon_position = pos2(icon_position_x, rect.center().y);
+        let icon_rect = Rect::from_center_size(icon_position, vec2(icon_size, icon_size));
 
         let painter = ui.painter();
 
         painter.add(Shape::Circle {
-            center: big_icon_rect.center(),
-            radius: big_icon_rect.width() / 2.0 + visuals.expansion,
+            center: icon_rect.center(),
+            radius: icon_size * 0.5,
             fill: visuals.bg_fill,
             stroke: visuals.bg_stroke,
         });
 
         if *checked {
+            let fill = if *color == Color32::TRANSPARENT {
+                ui.visuals().noninteractive().fg_stroke.color
+            } else {
+                *color
+            };
             painter.add(Shape::Circle {
-                center: small_icon_rect.center(),
-                radius: small_icon_rect.width() * 0.8,
-                fill: *color,
+                center: icon_rect.center(),
+                radius: icon_size * 0.4,
+                fill,
                 stroke: Default::default(),
             });
         }
 
-        let text_position = pos2(
-            rect.left() + padding.x + icon_width + icon_spacing,
-            rect.center().y - 0.5 * galley.size.y,
-        );
+        let text_position_x = if flipped {
+            rect.right() - icon_size - icon_spacing - galley.size.x
+        } else {
+            rect.left() + icon_size + icon_spacing
+        };
+
+        let text_position = pos2(text_position_x, rect.center().y - 0.5 * galley.size.y);
         painter.galley(text_position, galley, visuals.text_color());
 
-        self.checked ^= response.clicked_by(PointerButton::Primary);
-        self.hovered = response.hovered();
+        *checked ^= response.clicked_by(PointerButton::Primary);
+        *hovered = response.hovered();
 
         response
     }
 }
+
+#[derive(Clone)]
+pub(super) struct LegendWidget {
+    rect: Rect,
+    entries: BTreeMap<String, LegendEntry>,
+    config: Legend,
+}
+
+impl LegendWidget {
+    /// Create a new legend from items, the names of items that are hidden and the style of the
+    /// text. Returns `None` if the legend has no entries.
+    pub(super) fn try_new(
+        rect: Rect,
+        config: Legend,
+        items: &[Box<dyn PlotItem>],
+        hidden_items: &HashSet<String>,
```

Comment:
it feels a bit backwards to have a list of things to hide rather than a list of things to show

Reply:
Yes, but it seems to be the only way. I can't pass things that are shown between frames. Because if a new curve is added, it will not be shown because it's not already in the "shown curves".